pact-lsp/Pact/Core/LanguageServer.hs

```diff
@@ -319,6 +325,69 @@ documentHoverRequestHandler = requestHandler SMethod_TextDocumentHover $ \req re
         debug "documentHover: could not find term on position"
         resp (Right (InR Null))
 
+
+-- newtype Collect a = Collect { getCollect :: [a] }
+--   deriving (Eq, Show)
+
+-- instance Semigroup (Collect a) where
+--   Collect a <> Collect b = Collect (a ++ b)
+
+-- instance Monoid (Collect a) where
+--   mempty = Collect []
+
+-- collectIf :: Traversable t => t a -> (a -> Bool) -> [a]
+-- collectIf t p = getCollect (foldMap (\x -> if p x then Collect [x] else mempty) t)
+
+
+documentRenameRequestHandler :: Handlers LSM
+documentRenameRequestHandler = requestHandler SMethod_TextDocumentRename $ \req resp ->
+  getState >>= \st -> do
+    let
+        uri' = req ^. params . textDocument . uri
+        nuri = req ^. params . textDocument . uri . to toNormalizedUri
+        pos = req ^. params . position
+        nName = req ^. params . newName
+        tls = fromMaybe [] $ view (lsTopLevel . at nuri) st
+        toTextEdit r = TextEdit r nName
+    case getMatch pos tls of
+      Nothing -> do
+        debug "documentRenameRequestHandler: could not find term at position"
+        resp (Right (InR Null))
+      Just tm -> do
+        changes <- case tm of
+                     TermMatch (Var (Name n vt) _) -> case vt of
+                       NBound _db -> error "nbound"
+                       NTopLevel _mn _mh -> do
+                         let isSameVar = \case
+                               Var (Name n' vt') _
+                                 | n == n' && vt == vt' -> True
+                               _ -> False
+                             occurences = concatMap (matchingTerms isSameVar) tls
+
+                         forM occurences $ \case
+                           Var _ i -> pure $ toTextEdit (spanInfoToRange i)
+                           _ -> error "invariant"
```

Comment:
Probably still better to fail via whatever mechanism `Handlers` provides, with a message like "The impossible happened: ..."